### PR TITLE
#12253: Implement Batch norm operation for inference mode

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### List of Github issues

- https://github.com/tenstorrent/tt-metal/issues/12253 
- https://github.com/tenstorrent/tt-metal/issues/15449

### Problem description
To implement batch normalization as device op

### What's changed
Work done : 

- Provided support for Inference mode

Screenshot 2025-01-10 at 11 59 49 AM

### Checklist
- [x] [All post-commit testsI](https://github.com/tenstorrent/tt-metal/actions/runs/12729863177)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/12729863432)
- [ ] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) -  [Link to test](https://github.com/tenstorrent/tt-metal/actions/runs/12732676848)
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/12729864298/job/35486219235) - same as main
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/12729865036) - same as main
- [x] [Single-card Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/12729865358) - same as main

### CODE REVIEW

**Review by Aswinmcw — APPROVED** [review]
LGTM

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    auto C = input.get_shape().value[1];
+    // output (N, C, H, W)
+    if (output.has_value()) {
+        auto check_C = output.value().get_shape().value[1];
+        TT_FATAL(C == check_C, "output_shape[1] must be the same as input's channel size.");
+    }
+
+    // mean (1, C, 1, 1)
+    TT_FATAL(
+        batch_mean.get_shape().value.without_padding()[1] == C,
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:39]: get_logical_shape() maybe?

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    if (num_tiles == 0) {
+        return;
+    }
+
+    constexpr auto cb_input = tt::CBIndex::c_0;       // input
+    constexpr auto cb_batch_mean = tt::CBIndex::c_1;  // batch_mean
+    constexpr auto cb_output_0 =
+        tt::CBIndex::c_2;  // output -- > [(input - batch_mean)/(sqrt(batch_var + eps))] * weight
+    constexpr auto cb_batch_var = tt::CBIndex::c_3;  // batch_var
+    constexpr auto cb_eps = tt::CBIndex::c_4;        // batch_var
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp:52]: //

**Review by KalaivaniMCW — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        a_data_format);  // to store input - batch_mean
+    auto [temp_1_cb, temp_1_cb_handle] =
+        create_cb(tt::CBIndex::c_17, program, all_device_cores, a_single_tile_size, num_tiles_per_cb, a_data_format);
+
+    auto a_is_dram = static_cast<uint32_t>(a_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto b_is_dram = static_cast<uint32_t>(b_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto c_is_dram = static_cast<uint32_t>(c_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto d_is_dram = static_cast<uint32_t>(d_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    bool e_is_dram = false;
+    bool f_is_dram = false;
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp:232]:
```suggestion
    const auto e_is_dram = weight_has_value and e->buffer_type() == tt_metal::BufferType::DRAM;
    const auto f_is_dram = bias_has_value and f->buffer_type() == tt_metal::BufferType::DRAM;
```
and delete the declarations for `e_buffer` and `f_buffer` above, and the two respective conditional blocks below.

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        a_single_tile_size,
+        num_tiles_per_cb,
+        a_data_format);  // to store input - batch_mean
+    auto [temp_1_cb, temp_1_cb_handle] =
+        create_cb(tt::CBIndex::c_17, program, all_device_cores, a_single_tile_size, num_tiles_per_cb, a_data_format);
+
+    auto a_is_dram = static_cast<uint32_t>(a_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto b_is_dram = static_cast<uint32_t>(b_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto c_is_dram = static_cast<uint32_t>(c_buffer->buffer_type() == tt_metal::BufferType::DRAM);
+    auto d_is_dram = static_cast<uint32_t>(d_buffer->buffer_type() == tt_metal::BufferType::DRAM);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_program_factory.cpp:230]: Can we just use `a.buffer()->buffer_type() ...` etc. directly here? These are the only places the `*_buffer` variables are used, and we can delete the declarations above.

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+        tile_regs_wait();
+        pack_tile_with_dt(dst0, cb_affine_or_out);
+        tile_regs_release();
+
+        cb_pop_front(cb_num, 1);
+        cb_pop_front(cb_den, 1);
+        cb_push_back(cb_affine_or_out, onetile);
+
+        if (weight_has_value) {  // result = result * weight
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp:119]:
```suggestion
        if constexpr (weight_has_value) {  // result = result * weight
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+            tile_regs_wait();
+            pack_tile_with_dt(dst0, cb_scaled_output);
+            tile_regs_release();
+
+            cb_pop_front(cb_affine_or_out, 1);
+            cb_pop_front(cb_weight, 1);
+            cb_push_back(cb_scaled_output, onetile);
+        }
+        if (bias_has_value) {  // result = result + bias
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/batch_norm_kernel.cpp:137]:
```suggestion
        if constexpr (bias_has_value) {  // result = result + bias
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_batch_norm.cpp [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+            // read a tile from batch variance
+            cb_reserve_back(cb_id_batch_var, onetile);
+            uint32_t l1_batch_var_write_addr = get_write_ptr(cb_id_batch_var);
+            noc_async_read_tile(tile_offset, batch_var, l1_batch_var_write_addr);
+            noc_async_read_barrier();
+            fill_tile_with_first_element_bfloat16(cb_id_batch_var);
+            cb_push_back(cb_id_batch_var, onetile);
+
+            if (weight_has_value) {  // read a tile from weight tensor
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_batch_norm.cpp:102]:
```suggestion
            if constexpr (weight_has_value) {  // read a tile from weight tensor
```

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_batch_norm.cpp [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                cb_reserve_back(cb_id_weight, onetile);
+                uint32_t l1_weight_write_addr = get_write_ptr(cb_id_weight);
+                noc_async_read_tile(tile_offset, weight, l1_weight_write_addr);
+                noc_async_read_barrier();
+                fill_tile_with_first_element_bfloat16(cb_id_weight);
+                cb_push_back(cb_id_weight, onetile);
+            }
+
+            if (bias_has_value) {  // read a tile from bias tensor
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_batch_norm.cpp:111]:
```suggestion
            if constexpr (bias_has_value) {  // read a tile from bias tensor
```

**Review by patrickroberts — CHANGES_REQUESTED** [review]

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+    )
+    var_data, var_tensor = data_gen_with_range_batch_norm(
+        input_shapes, 4, 20, device, is_affine=False, is_mean_var=True, is_training=training
+    )
+    weight_data, weight_tensor = data_gen_with_range_batch_norm(
+        input_shapes, 4, 10, device, is_affine=True, weight=weight, is_mean_var=False, is_training=training
+    )
+    bias_data, bias_tensor = data_gen_with_range_batch_norm(
+        input_shapes, 4, 10, device, is_affine=True, bias=bias, is_mean_var=False, is_training=training
+    )
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:72]:
```suggestion
    in_data, input_tensor = data_gen_with_range_batch_norm(
        input_shapes, 5, 10, device, is_input=True
    )
    mean_data, mean_tensor = data_gen_with_range_batch_norm(
        input_shapes, 4, 10, device
    )
    var_data, var_tensor = data_gen_with_range_batch_norm(
        input_shapes, 4, 20, device
    )
    weight_data, weight_tensor = data_gen_with_range_batch_norm(
        input_shapes, 4, 10, device
    ) if weight else (None, None)
    bias_data, bias_tensor = data_gen_with_range_batch_norm(
        input_shapes, 4, 10, device
    ) if bias else (None, None)
```

**Review by patrickroberts — CHANGES_REQUESTED** [review]

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+        (torch.Size([1, 2, 64, 120])),
+        (torch.Size([1, 3, 64, 120])),
+        (torch.Size([2, 1, 64, 120])),
+        (torch.Size([2, 2, 64, 120])),
+        (torch.Size([2, 3, 64, 120])),
+        (torch.Size([3, 1, 64, 120])),
+        (torch.Size([3, 2, 64, 120])),
+    ),
+)
+@pytest.mark.parametrize("training", [False])
```
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/test_batch_norm.py:53]: In the upcoming PR, We will be testing for True test case as well. That should be up by this week.

File: tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py [hunk 11]
```diff
@@ -11,6 +11,42 @@
...
+        )
+    elif ((not is_training) and is_mean_var) or (is_affine and (weight or bias)):
+        pt_tensor = torch.rand(channels, requires_grad=required_grad).bfloat16() * (high - low) + low
+        # pt_tensor = pt_tensor.view(1, channels, 1, 1) # to test each section of TT op
+        reshaped_tensor = pt_tensor.view(1, channels, 1, 1).expand(1, channels, 32, 32)
+        tt_tensor = ttnn.Tensor(reshaped_tensor, ttnn.bfloat16).to(ttnn.TILE_LAYOUT).to(device)
+    else:
+        pt_tensor = None
+        tt_tensor = None
+    return pt_tensor, tt_tensor
```
[the remark this answers is not in the document]
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py:47]: I would need reshaped tensor for torch input. Will update accordingly

**Review by patrickroberts — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py [hunk 12]
```diff
@@ -11,6 +11,32 @@
...
+    required_grad=False,
+):
+    assert high > low, "Incorrect range provided"
+    torch.manual_seed(213919)
+    channels = input_shapes[1]
+    size = input_shapes if is_input else channels
+    pt_tensor = torch.rand(size, requires_grad=required_grad).bfloat16() * (high - low) + low
+    reshaped_tensor = pt_tensor
+    if not is_input:
+        reshaped_tensor = pt_tensor.view(1, channels, 1, 1).expand(1, channels, 32, 32)
```
Comment (patrickroberts) [inline tests/ttnn/unit_tests/operations/eltwise/backward/utility_funcs.py:29]:
```suggestion
    reshaped_tensor = pt_tensor if is_input else pt_tensor.view(1, channels, 1, 1).expand(1, channels, 32, 32)
```
I think this would make it a lot less confusing to follow. I thought `reshaped_tensor` was unnecessary until I noticed `pt_tensor` in the return statement.

**Review by dmakoviichuk-tt — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_batch_norm.py [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        *(torch.Size([n, c, 64, 120]) for n, c in product([1, 2], [1, 2, 3])),
+        torch.Size([3, 1, 64, 120]),
+        torch.Size([3, 2, 64, 120]),
+    ],
+)
+@pytest.mark.parametrize("training", [False])
+@pytest.mark.parametrize("weight", [True, False])
+@pytest.mark.parametrize("bias", [True, False])
+@pytest.mark.parametrize("eps", [1.0, 0.0, 2.34, 1e-05])
+def test_batch_norm(input_shapes, training, weight, bias, eps, device):
```
Comment (yugaoTT) [inline tests/ttnn/unit_tests/operations/test_batch_norm.py:31]: can you also add a test to test program_cache?

**Review by yugaoTT — APPROVED** [review]

**Review by eyonland — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm_pybind.cpp [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+#include "pybind11/decorators.hpp"
+namespace py = pybind11;
+namespace ttnn::operations::normalization::detail {
+void bind_batch_norm_operation(pybind11::module& module) {
+    ttnn::bind_registered_operation(
+        module,
+        ttnn::batch_norm,
+        R"doc(
+            Applies Spatial Batch Normalization over each channel on :attr:`input_tensor`.Currently support is provided for inference mode only.
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm_pybind.cpp:17]: You're missing a space after a period. Should be ". Currently"

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+BatchNormOperation::program_factory_t BatchNormOperation::select_program_factory(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    return BatchNormFactory();
+}
+
+void BatchNormOperation::validate_on_program_cache_miss(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    // We don't support sharding for now
```
Comment (bbradelTT) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/batch_norm_device_operation.cpp:57]: The limitations should be listed in the pydoc string.

**Review by bbradelTT — APPROVED** [review]

File: tests/ttnn/unit_tests/operations/test_batch_norm.py (hunk 13 above)
Comment (VirdhatchaniKN) [reply tests/ttnn/unit_tests/operations/test_batch_norm.py:31]: I've added that now

**Review by mouliraj-mcw — APPROVED** [review]

**Review by umadevimcw — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/12732681117 - passed

**Comment by bbradelTT** [comment]
@VirdhatchaniKN does this PR also fix https://github.com/tenstorrent/tt-metal/issues/15372 ?

**Comment by VirdhatchaniKN** [comment]
> @VirdhatchaniKN does this PR also fix #15372 ?

Yes @bbradelTT
